src/pgwire/src/protocol.rs

```diff
@@ -236,30 +237,43 @@ where
         let _ = session.vars_mut().set(&name, &value, local);
     }
 
-    // Register session with adapter.
-    let (mut adapter_client, startup) =
-        match adapter_client.startup(session, frontegg.is_some()).await {
-            Ok(startup) => startup,
-            Err(e) => {
-                return conn
-                    .send(ErrorResponse::from_adapter_error(Severity::Fatal, e))
-                    .await
-            }
-        };
+    // Register session with adapter. Use delayed startup so we can immediately
+    // respond as connected without waiting on the coordinator.
+    let delayed = adapter_client.startup_delayed(session, frontegg.is_some());
 
-    let session = adapter_client.session();
     let mut buf = vec![BackendMessage::AuthenticationOk];
+    buf.push(BackendMessage::BackendKeyData {
+        conn_id: conn.id(),
+        secret_key: delayed.secret_key,
+    });
+    buf.push(BackendMessage::ReadyForQuery(TransactionCode::Idle));
+    conn.send_all(buf).await?;
+    conn.flush().await?;
+
+    // Now that we've issued a connected status to the client, wait for the
+    // coordinator to respond about if we actually succeeded or not.
+    let (mut adapter_client, response) = delayed.response_fut.await;
+    let startup = match response {
+        Ok(startup) => startup,
+        Err(e) => {
+            adapter_client.take_session();
+            return conn
+                .send(ErrorResponse::from_adapter_error(Severity::Fatal, e))
+                .await;
+        }
+    };
+    let session = adapter_client.session();
+
+    // ParameterStatus and NoticeResponse messages can be sent at any time
+    // (https://www.postgresql.org/docs/current/protocol-flow.html#PROTOCOL-ASYNC),
+    // so it is within spec to send them after the initial ReadyForQuery.
+    let mut buf = Vec::new();
     for var in session.vars().notify_set() {
```

Comment:
Perhaps the session vars can be pre-generated on the assumption that Startup doesn't change any of them? Then only the notice messages get delayed.

Reply:
Yes, that sounds right! If the coordinator wants to arbitrarily change session values outside of a `SetVariable` command we'll have to invent some new protocol for notifying the session of the changed value anyway; so it seems totally reasonable to send the parameter status messages before the startup future resolves.